File: c++_solutions/41.cpp
// hint : try to mark the element in range (1 to n)
// hint : for marking we can use the array index 
// hint : whenever we find that a number is not on correct position we can place that number on correct postion by swapping the element on current position with the element on the position denoted by the current element (so in the end we will have that number on its correct position)
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size() ; 
        for(size_t i = 0 ; i < nums.size() ; i++){
            while(nums[i] >= 1 && nums[i] <= n && nums[i] != nums[nums[i]-1]){
                swap(nums[i] , nums[nums[i]-1]) ;
            }
        }
        for(int i  = 0 ; i < n; i++){
            if(nums[i]  != i+1){
                return i+1 ; 
            }
        }
        return n+1 ;
    }
};

File: python_solutions/41.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        curr_idx = 0 
        while curr_idx < len(nums):
            while 0 <= nums[curr_idx]-1 < len(nums) and nums[curr_idx] != nums[nums[curr_idx]-1]:
                tmp_num = nums[curr_idx]
                nums[curr_idx] = nums[tmp_num-1]
                nums[tmp_num-1] = tmp_num
            curr_idx += 1
        curr_idx = 0
        while curr_idx < len(nums):
            if curr_idx+1 != nums[curr_idx]:
                return curr_idx+1
            curr_idx += 1
        return curr_idx+1
